src/analysis/normed_space/lp_space.lean

```diff
@@ -628,6 +628,99 @@ end
 
 end normed_space
 
+section non_unital_normed_ring
+
+variables {I : Type*} {B : I → Type*} [Π i, non_unital_normed_ring (B i)]
+
+lemma _root_.mem_ℓp.infty_mul {f g : Π i, B i} (hf : mem_ℓp f ∞) (hg : mem_ℓp g ∞) :
+  mem_ℓp (f * g) ∞ :=
+begin
+  rw mem_ℓp_infty_iff,
+  obtain ⟨⟨Cf, hCf⟩, ⟨Cg, hCg⟩⟩ := ⟨hf.bdd_above, hg.bdd_above⟩,
+  refine ⟨Cf * Cg, _⟩,
+  rintros _ ⟨i, rfl⟩,
+  calc ∥(f * g) i∥ ≤ ∥f i∥ * ∥g i∥ : norm_mul_le (f i) (g i)
+  ...             ≤ Cf * Cg       : mul_le_mul (hCf ⟨i, rfl⟩) (hCg ⟨i, rfl⟩) (norm_nonneg _)
+                                      ((norm_nonneg _).trans (hCf ⟨i, rfl⟩))
+end
+
+instance : has_mul (lp B ∞) :=
+{ mul := λ f g, ⟨(f  * g : Π i, B i) , f.property.infty_mul g.property⟩}
+
+@[simp] lemma infty_coe_fn_mul (f g : lp B ∞) : ⇑(f * g) = f * g := rfl
+
+
+
+instance : non_unital_ring (lp B ∞) :=
+function.injective.non_unital_ring lp.has_coe_to_fun.coe (by ext) (lp.coe_fn_zero B ∞)
+  lp.coe_fn_add infty_coe_fn_mul lp.coe_fn_neg lp.coe_fn_sub (λ _ _, rfl) (λ _ _,rfl)
+
+instance : non_unital_normed_ring (lp B ∞) :=
+{ norm_mul := λ f g, lp.norm_le_of_forall_le (mul_nonneg (norm_nonneg f) (norm_nonneg g))
+    (λ i, calc ∥(f * g) i∥ ≤ ∥f i∥ * ∥g i∥ : norm_mul_le _ _
+    ...                    ≤ ∥f∥ * ∥g∥
+    : mul_le_mul (lp.norm_apply_le_norm ennreal.top_ne_zero f i)
+        (lp.norm_apply_le_norm ennreal.top_ne_zero g i) (norm_nonneg _) (norm_nonneg _)),
+  .. lp.normed_group }
+
+-- we also want a `non_unital_normed_comm_ring` instance, but this has to wait for #13719
+
+end non_unital_normed_ring
+
+section normed_ring
+
+variables  {I : Type*} {B : I → Type*} [Π i, normed_ring (B i)] [Π i, norm_one_class (B i)]
+
+lemma _root_.mem_ℓp.infty_one : mem_ℓp (1 : Π i, B i) ∞ :=
+⟨1, by { rintros i ⟨i, rfl⟩, exact norm_one.le,}⟩
+
+instance : has_one (lp B ∞) :=
+{ one := ⟨(1 : Π i, B i), mem_ℓp.infty_one⟩ }
+
+@[simp] lemma infty_coe_fn_one : ⇑(1 : lp B ∞) = 1 := rfl
+
+lemma _root_.mem_ℓp.infty_pow {f : Π i, B i} (hf : mem_ℓp f ∞) (n : ℕ) : mem_ℓp (f ^ n) ∞ :=
+begin
+  induction n with n hn,
+  rw pow_zero,
+  exact mem_ℓp.infty_one,
+  rw pow_succ,
+  exact hf.infty_mul hn,
+end
+
+instance [nonempty I] : norm_one_class (lp B ∞) :=
+begin
+  split,
+  rw lp.norm_eq_csupr,
+  convert csupr_const,
+  simp only [infty_coe_fn_one, pi.one_apply, norm_one],
+  apply_instance
+end
+
+instance : ring (lp B ∞) :=
+{ one_mul := λ f, by { ext1, rw lp.infty_coe_fn_mul, simp only [infty_coe_fn_one, one_mul] },
+  mul_one := λ f, by { ext1, rw lp.infty_coe_fn_mul, simp only [infty_coe_fn_one, mul_one] },
+  .. lp.non_unital_ring,
+  .. lp.has_one }
```

Comment:
Here's the `function.injective version`:
```suggestion
instance : has_pow (lp B ∞) ℕ := { pow := λ f n, ⟨_, f.prop.infty_pow n⟩ }

@[simp] lemma infty_coe_fn_pow (f : lp B ∞) (n : ℕ) : ⇑(f ^ n) = f ^ n := rfl

instance : ring (lp B ∞) :=
function.injective.ring lp.has_coe_to_fun.coe subtype.coe_injective
  (lp.coe_fn_zero B ∞) (infty_coe_fn_one) lp.coe_fn_add infty_coe_fn_mul
  lp.coe_fn_neg lp.coe_fn_sub (λ _ _, rfl) (λ _ _, rfl) infty_coe_fn_pow
```

It's longer, but it gives us a free proof of `⇑(f ^ n) = f ^ n`, which your version doesn't.

Reply:
why does this happen?

Reply:
My computer seem to think the proof of infty_coe_fn_pow is free (if by free you mean provable by rfl).

Reply:
oh I think I see what you getting at, in order to do ring your way you have to do the pow stuff so it ends longer than the way it is currently being done, however you probs want to done the pow stuff anyway so it doesn't really matter.